### DISCUSSION TITLE
Run a lean VM without containerd/nerdctl

### DISCUSSION THREAD

**Discussion by chenxiaoqino** [lima-vm/lima#631] [Q&A] [answered]
### Description

I am trying to run a lightweight VM on hardware-constrained laptops, and want to minimize the VM's unnecessary CPU/Memory footprint. In my use case, I run single pre-built Linux binaries in the VM, and do not need containerd/nerdctl. However, a bare-bone default VM with nothing running costs 5-10% CPU and close to 1GB host memory (~300MB resident as seen inside the VM), mostly due to many containerd-related processes.

Since I don't run containers I want to remove these background processes in the VM. However, seems like there is no option to avoid adding containerd-related stuff when launching the VM. The launch script always attempt to download and install nerdctl -- it appears that even if I delete everything, it gets installed again in the next reboot. (I believe the "sticky" logic is here: https://github.com/lima-vm/lima/blob/67e465e2a57ad673cb41e5c8b51695dc783e401b/pkg/cidata/cidata.TEMPLATE.d/boot/40-install-containerd.sh#L11)

I can open a pull-request myself if the authors can advise on where to change the logic. Is [start.go](https://github.com/lima-vm/lima/blob/master/pkg/start/start.go) the best place to change?

**Comment by jandubois** [comment, marked as answer]
> However, seems like there is no option to avoid adding containerd-related stuff when launching the VM.

You can disable installation of nerdctl and containerd by editing your `lima.yaml` file:

https://github.com/lima-vm/lima/blob/f24302b6d0a8b5ea2666a6b5eea169d64c20ccea/pkg/limayaml/default.yaml#L90-L92

The default for `containerd.user` is `true`, so set it to `false` to disable:

```yaml
containerd:
  system: false
  user: false
```

This is done in the [Alpine example](https://github.com/lima-vm/lima/blob/master/examples/alpine.yaml), which is also generally the smallest instance you can use.

**Reply by chenxiaoqino** [reply]
Thanks a lot!

**Reply by afbjorklund** [reply]
> This is done in the [Alpine example](https://github.com/lima-vm/lima/blob/master/examples/alpine.yaml)

But that's probably a "bug", right ?

Once the nerdctl installation works on Alpine, it should probably be included there - like with any other distro ?

* https://github.com/lima-vm/lima/issues/310

The recommendation to set them both to false to _avoid_ installing nerdctl/containerd/buildkit still holds, though.

**Reply by jandubois** [reply]
> But that's probably a "bug", right ?

It is not a bug, it is a limitation! 😛
 
> Once the nerdctl installation works on Alpine, it should probably be included there - like with any other distro ?

It probably should, but not totally sure if we should still change it after it being a bare/minimal instance for such a long time. Consistency vs. expectations... I guess I could be persuaded either way still...

**Reply by afbjorklund** [reply]
Oh well, it is spelled out in the yaml so should not come as a surprise either way.
